packages/eslint-plugin/src/rules/strict-enums.ts

```diff
@@ -0,0 +1,806 @@
+          util.isTypeFlagSet(
+            parameterSubType,
+            ALLOWED_TYPES_FOR_ANY_ENUM_ARGUMENT,
+          )
+        ) {
+          return false;
+        }
+      }
+
+      /**
+       * Disallow passing number literals or string literals into functions that
+       * take in an enum, like the following:
+       *
+       * ```ts
+       * function useFruit(fruit: Fruit) {}
+       * declare const fruit: Fruit.Apple | 1;
+       * useFruit(fruit)
+       * ```
+       */
+      const argumentSubTypes = tsutils.unionTypeParts(argumentType);
+      for (const argumentSubType of argumentSubTypes) {
+        if (argumentSubType.isLiteral() && !isEnum(argumentSubType)) {
+          return true;
+        }
+      }
+
+      /**
+       * Allow function calls that match one of the types in a union, like the
+       * following:
+       *
+       * ```ts
+       * function useApple(fruitOrNull: Fruit | null) {}
+       * useApple(null);
+       * ```
+       */
+      const argumentSubTypesSet = new Set(argumentSubTypes);
+      const parameterSubTypesSet = new Set(parameterSubTypes);
+      if (setHasAnyElementFromSet(argumentSubTypesSet, parameterSubTypesSet)) {
+        return false;
+      }
+
+      /**
+       * Allow function calls that have a base enum that match the function
+       * type, like the following:
+       *
+       * ```ts
+       * function useFruit(fruit: Fruit) {}
+       * useFruit(Fruit.Apple);
+       * ```
+       */
+      if (setHasAnyElementFromSet(argumentEnumTypes, parameterEnumTypes)) {
+        return false;
+      }
+
+      return true;
+    }
+
+    // ------------------
+    // AST node callbacks
+    // ------------------
+
+    return {
+      /** When something is assigned to a variable. */
+      AssignmentExpression(node): void {
+        const leftType = getTypeFromNode(node.left);
+        const rightType = getTypeFromNode(node.right);
+
+        if (isAssigningNonEnumValueToEnumVariable(leftType, rightType)) {
+          context.report({
+            node,
+            messageId: 'mismatchedAssignment',
+          });
+        }
+      },
+
+      /** When a comparison between two things happen. */
+      BinaryExpression(node): void {
+        const leftType = getTypeFromNode(node.left);
+        const rightType = getTypeFromNode(node.right);
+
+        if (isMismatchedEnumComparison(node.operator, leftType, rightType)) {
+          context.report({
+            node,
+            messageId: 'mismatchedComparison',
+          });
+        }
+      },
+
+      /** When a function is invoked. */
+      CallExpression(node): void {
+        checkCallExpression(node);
+      },
+
+      /** When something is instantiated with the "new" keyword. */
+      NewExpression(node): void {
+        /**
+         * We need to perform the exact same checks on a class constructor
+         * invocation as we do on a normal function invocation.
+         */
+        checkCallExpression(node);
+      },
```

Comment:
nit can combine these
```suggestion
      /** When a function is invoked, or class is instantiated. */
      'CallExpression, NewExpression'(
        node: TSESTree.CallExpression | TSESTree.NewExpression,
      ): void {
        checkCallExpression(node);
      },
```